File: 2025/07 July/27 Set Matrix Zeros/main.cpp
// Link: https://www.geeksforgeeks.org/problems/set-matrix-zeroes/1

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    void setMatrixZeroes(vector<vector<int>> &mat)
    {
        int n = mat.size(), m = mat[0].size();

        int c0 = 1;

        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                if (mat[i][j] == 0)
                {
                    mat[i][0] = 0;
                    if (j == 0)
                        c0 = 0;
                    else
                        mat[0][j] = 0;
                }
            }
        }

        for (int i = 1; i < n; i++)
        {
            for (int j = 1; j < m; j++)
            {
                if (mat[i][0] == 0 || mat[0][j] == 0)
                {
                    mat[i][j] = 0;
                }
            }
        }

        if (mat[0][0] == 0)
        {
            for (int j = 0; j < m; j++)
                mat[0][j] = 0;
        }

        if (c0 == 0)
        {
            for (int i = 0; i < n; i++)
                mat[i][0] = 0;
        }
    }
};

int main()
{
    Solution sol;
    vector<vector<int>> mat = {
        {1, -1, 1},
        {-1, 0, 1},
        {1, -1, 1}};
    sol.setMatrixZeroes(mat);
    cout << "Matrix after setting zeroes:" << endl;
    for (const auto &row : mat)
    {
        for (int val : row)
        {
            cout << val << " ";
        }
        cout << endl;
    }
    cout << "Done!" << endl;

    mat = {
        {0, 1, 2, 0},
        {1, 2, 6, 2},
        {1, 3, 1, 8}};
    sol.setMatrixZeroes(mat);
    cout << "Matrix after setting zeroes:" << endl;
    for (const auto &row : mat)
    {
        for (int val : row)
        {
            cout << val << " ";
        }
        cout << endl;
    }
    cout << "Done!" << endl;

    return 0;
}

File: 2025/07 July/27 Set Matrix Zeros/main.py
# Link: https://www.geeksforgeeks.org/problems/set-matrix-zeroes/1


class Solution:
    def setMatrixZeroes(self, mat):
        n = len(mat)
        m = len(mat[0])

        c0 = 1

        for i in range(n):
            for j in range(m):
                if mat[i][j] == 0:
                    mat[i][0] = 0

                    if j == 0:
                        c0 = 0
                    else:
                        mat[0][j] = 0

        for i in range(1, n):
            for j in range(1, m):
                if mat[i][0] == 0 or mat[0][j] == 0:
                    mat[i][j] = 0

        if mat[0][0] == 0:
            for j in range(m):
                mat[0][j] = 0

        if c0 == 0:
            for i in range(n):
                mat[i][0] = 0


if __name__ == "__main__":
    sol = Solution()
    mat = [[1, -1, 1], [-1, 0, 1], [1, -1, 1]]
    sol.setMatrixZeroes(mat)
    print("Matrix after setting zeroes:")
    for row in mat:
        print(row)
    print("Done!")

    mat = [[0, 1, 2, 0], [1, 2, 6, 2], [1, 3, 1, 8]]
    sol.setMatrixZeroes(mat)
    print("Matrix after setting zeroes:")
    for row in mat:
        print(row)
    print("Done!")
